src/core/lib/event_engine/windows/windows_endpoint.cc

```diff
@@ -68,76 +76,79 @@ WindowsEndpoint::~WindowsEndpoint() {
-  // Did we get data immediately ? Yay.
   if (wsa_error != WSAEWOULDBLOCK) {
+    // Data or some error was returned immediately.
     io_state_->socket->read_info()->SetResult(
         {/*wsa_error=*/wsa_error, /*bytes_read=*/bytes_read});
-    absl::Status result;
-    if (bytes_read == 0) {
-      result = absl::UnavailableError("End of TCP stream");
-      grpc_core::StatusSetInt(&result, grpc_core::StatusIntProperty::kRpcStatus,
-                              GRPC_STATUS_UNAVAILABLE);
-      buffer->Clear();
-    } else {
-      result = absl::OkStatus();
-      // prune slicebuffer
-      if (bytes_read != buffer->Length()) {
-        buffer->RemoveLastNBytes(buffer->Length() - bytes_read);
-      }
-    }
-    executor_->Run(
-        [result, on_read = std::move(on_read)]() mutable { on_read(result); });
-    return false;
+    executor_->Run(&io_state_->handle_read_event);
+    return absl::OkStatus();
+  }
+  // If the endpoint has already received some data, and the next call would
+  // block, return the data in case that is all the data the reader expects.
+  if (io_state_->handle_read_event.MaybeFinishIfDataHasAlreadyBeenRead(
+          absl::OkStatus())) {
+    return absl::OkStatus();
   }
   // Otherwise, let's retry, by queuing a read.
-  memset(io_state_->socket->read_info()->overlapped(), 0, sizeof(OVERLAPPED));
   status = WSARecv(io_state_->socket->raw_socket(), wsa_buffers,
                    (DWORD)buffer->Count(), &bytes_read, &flags,
                    io_state_->socket->read_info()->overlapped(), nullptr);
   wsa_error = status == 0 ? 0 : WSAGetLastError();
   if (wsa_error != 0 && wsa_error != WSA_IO_PENDING) {
     // Async read returned immediately with an error
-    executor_->Run([this, on_read = std::move(on_read), wsa_error]() mutable {
-      on_read(GRPC_WSA_ERROR(
-          wsa_error,
-          absl::StrFormat("WindowsEndpont::%p Read failed", this).c_str()));
+    return GRPC_WSA_ERROR(
+        wsa_error,
+        absl::StrFormat("WindowsEndpont::%p Read failed", this).c_str());
+  }
+  io_state_->socket->NotifyOnRead(&io_state_->handle_read_event);
+  return absl::OkStatus();
+}
+
+bool WindowsEndpoint::Read(absl::AnyInvocable<void(absl::Status)> on_read,
+                           SliceBuffer* buffer, const ReadArgs* /* args */) {
+  GRPC_EVENT_ENGINE_ENDPOINT_TRACE("WindowsEndpoint::%p reading", this);
+  if (io_state_->socket->IsShutdown()) {
+    executor_->Run([on_read = std::move(on_read)]() mutable {
+      on_read(absl::UnavailableError("Socket is shutting down."));
     });
     return false;
   }
+  buffer->Clear();
+  io_state_->handle_read_event.DonateSpareSlices(buffer);
+  // TODO(hork): sometimes args->read_hint_bytes is 1, which is not useful.
+  // Choose an appropriate size.
+  size_t min_read_size = kDefaultTargetReadSize;
+  if (buffer->Length() < min_read_size && buffer->Count() < kMaxWSABUFCount) {
+    buffer->AppendIndexed(Slice(allocator_.MakeSlice(min_read_size)));
```

Comment:
I'm not sure if it should, honestly. As it's currently written, the buffer size could end up being between `min_read_size` and `(2 * min_read_size) - 1`. If we only allocate up to `min_read_size`, we may end up doing more allocations over the life of the process because we've only allocated the minimum. It is a tradeoff between memory and cpu utilization, especially in applications with a lot of open endpoints. I'd wager that there aren't many low-memory Windows systems that gRPC is running on, so I'd opt for a larger memory footprint here.

Reply:
Gotcha, interesting. Thanks for the explanation!